---
title: "isoformswitchanalyzer"
output: html_document
---

documentation here: https://bioconductor.org/packages/release/bioc/vignettes/IsoformSwitchAnalyzeR/inst/doc/IsoformSwitchAnalyzeR.html#overview-of-alternative-splicing-workflow
```{r}
library(IsoformSwitchAnalyzeR)
library(tidyverse)
```

import data
```{r}
kallistoQuant <- importIsoformExpression(parentDir = "kallisto/for_real_output")
myDesign <- data.frame(sampleID =  c("D6K20d2_191011_NextSeq.fastq.gz", "D6K20d3_191011_NextSeq.fastq.gz",  "D6K20W1_191011_NextSeq.fastq.gz", "D6Krw1_191011_NextSeq.fastq.gz", "D6Krw3_191011_NextSeq.fastq.gz", "D6L20d1_190919_NextSeq.fastq.gz", "D6L20d2_190919_NextSeq.fastq.gz", "D6L20d3_190919_NextSeq.fastq.gz", "D6L20W1_190903_NextSeq_R1.fastq.gz", "D6L20W2_190903_NextSeq_R1.fastq.gz", "D6L20W3_190903_NextSeq_R1.fastq.gz", "D6L40d1_190903_NextSeq_R1.fastq.gz", "D6L40d2_190903_NextSeq_R1.fastq.gz", "D6L40d3_190919_NextSeq.fastq.gz", "D6L40W1_190903_NextSeq_R1.fastq.gz", "D6L40W2_190903_NextSeq_R1.fastq.gz", "D6L40W3_190903_NextSeq_R1.fastq.gz", "D6L60d1_190903_NextSeq_R1.fastq.gz", "D6L60d2_190903_NextSeq_R1.fastq.gz", "D6L60d3_190903_NextSeq_R1.fastq.gz", "D6L60W1_190903_NextSeq_R1.fastq.gz", "D6L60W2_190903_NextSeq_R1.fastq.gz", "D6L60W3_190903_NextSeq_R1.fastq.gz", "D6Lrw1_190919_NextSeq.fastq.gz", "D6Lrw2_190919_NextSeq.fastq.gz", "D6Lrw3_190919_NextSeq.fastq.gz", "D7K20d2_191011_NextSeq.fastq.gz", "D7K20d5_191011_NextSeq.fastq.gz", "D7K20W1_191011_NextSeq.fastq.gz", "D7K20W2_191011_NextSeq.fastq.gz", "D7K40d1_191011_NextSeq.fastq.gz", "D7K40d2_191011_NextSeq.fastq.gz", "D7K40W2_190919_NextSeq.fastq.gz", "D7K40W3_191011_NextSeq.fastq.gz", "D7K40W4_191011_NextSeq.fastq.gz", "D7K60d1_190919_NextSeq.fastq.gz", "D7K60d2_190919_NextSeq.fastq.gz", "D7K60d3_191011_NextSeq.fastq.gz", "D7K60W2_190919_NextSeq.fastq.gz", "D7K60W3_190919_NextSeq.fastq.gz", "D7K60W4_190919_NextSeq.fastq.gz",  "D7Krw2_190919_NextSeq.fastq.gz"), condition = rep(c("K20d", "K20w", "Krw", "L20d", "L20w", "L40d", "L40w", "L60d", "L60w", "Lrw", "K20d", "K20w", "K40d", "K40w", "K60d", "K60w", "Krw"), times = c(2, 1, 2, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 3, 3, 3, 1)))
myDesign$sampleID <- as.factor(myDesign$sampleID)
myDesign$condition <- as.factor(myDesign$condition)
aSwitchList <- importRdata(isoformCountMatrix = kallistoQuant$counts, isoformRepExpression = kallistoQuant$abundance, designMatrix = myDesign, isoformExonAnnoation = "kallisto/AtRTD2_QUASI.gtf", isoformNtFasta = "kallisto/AtRTD2_QUASI.fasta")
#summary(aSwitchList)

#remove non-expressed isoforms, or genes with only 1 isoform (ie uninteresting things)
aSwitchList <- preFilter(aSwitchList)

#identify differentially used isoforms
#i suspect this test will be overly conservative, since it's doing every pairwise comparison of treatment groups. have to see what to do about that...
#be very careful with this, it takes 211 min to run -_-
aSwitchList <- isoformSwitchTestDEXSeq(aSwitchList, reduceToSwitchingGenes = F)

#we're missing some steps here, which is hindering what we can actually visualize. need to add some more functions here...

#this is broken cause the gtf don't have cds yet...
#aSwitchList <- addORFfromGTF(aSwitchList, pathToGTF = "kallisto/AtRTD2_QUASI.gtf")
#convert gtf to gff3, add cds to exons, then convert back -_-
#analyzeNovelOsoformORF might alos be a ood bet, since we have mostly novel isoforms here
#perhaps before we do this, we should add ORFs for all the standard transcripts from the ensembl GTF, thouhg idk if they'd have the same names...
#plus it seems like we can make our own pfam figures anyway...

#AS analysis
aSwitchList <- analyzeAlternativeSplicing(aSwitchList)
```

data preprocessing done. Should note that the chunk above takes 5ish hours to run locally. would be a good idea to run this overnight to work on the next day when needed. 
```{r}
#bar graph of AS events across comparisons
#this takes ~1h to run
pdf(file = 'extractSplicingSummary_all.pdf', onefile = FALSE, height=100, width = 50)
extractSplicingSummary(aSwitchList, asFractionTotal = F, plotGenes = F)
dev.off()
```
ok, so this works, but as expected, the table is a little too large to be useful. let's try subsetting, and pulling out only events of interest.

```{r}
smallSwitchList <- subsetSwitchAnalyzeRlist(aSwitchList, aSwitchList$isoformFeatures$condition_1 == "K20d")
```


```{r}
pdf(file = 'extractSplicingSummary_K20d.pdf', onefile = FALSE, height=100, width = 50)
extractSplicingSummary(smallSwitchList, asFractionTotal = F, plotGenes = F, splicingToAnalyze = c("A3", "A5", "ES", "IR"), localTheme = theme(text = element_text(size = 80)))
dev.off()
```

let's see if SBE2.1 shows differential isoform expression in K20d vs K20w
```{r}
switchPlot(
    smallSwitchList,
    gene='AT2G36390',
    condition1 = 'K20d',
    condition2 = 'K20w',
    localTheme = theme_bw(base_size = 60)
)
```
perhaps not... so what are the top genes then?

```{r}
k20dk20w <- subset(
    extractTopSwitches(
        smallSwitchList,
        n = 10000000,
        inEachComparison = F,
    )[,c('gene_id','condition_1','condition_2','gene_switch_q_value','Rank')]
)

k20dk20w <- k20dk20w %>% filter(condition_2 == "K20w")
#392 genes show significant isoform switching
#At2G36390 not in 'em
#write.table(k20dk20w, "dAS_analysis/dAS_K20dK20w.txt", col.names = T, row.names = F, sep = "\t", quote = F)
```

```{r}
switchPlotTranscript(smallSwitchList, gene = 'AT2G36390', condition1 = "K20d", condition2 = "K20w")
```


```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT2G36390', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT2G36390', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT2G36390', condition1 = "K20d", condition2 = "K20w")
```

let's look at PPL1 (PSBP-like protein 1)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT3G55330', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT3G55330', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT3G55330', condition1 = "K20d", condition2 = "K20w")
```

AT1G22850
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT1G22850', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT1G22850', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT1G22850', condition1 = "K20d", condition2 = "K20w")
```

AT3G23400 (FIB4)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT3G23400', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT3G23400', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT3G23400', condition1 = "K20d", condition2 = "K20w")
```

AT5G22640(EMB1211)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT5G22640', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT5G22640', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT5G22640', condition1 = "K20d", condition2 = "K20w")
```

AT4G24750 (STR11)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT4G24750', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT4G24750', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT4G24750', condition1 = "K20d", condition2 = "K20w")
```

AT3G44990 (XTH31)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT3G44990', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT3G44990', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT3G44990', condition1 = "K20d", condition2 = "K20w")
```

AT3G48570 (SEC61G3)
```{r}
switchPlotGeneExp(smallSwitchList, gene = 'AT3G48570', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoExp(smallSwitchList, gene = 'AT3G48570', condition1 = "K20d", condition2 = "K20w")
switchPlotIsoUsage(smallSwitchList, gene = 'AT3G48570', condition1 = "K20d", condition2 = "K20w")
```

```{r}
#i hate the plots this package is making, so imma make my own. write a ggplot function and save it nice and call it later. then we can make pretty figures for anything shared or presented.
smallSwitchList$isoformFeatures
test <- smallSwitchList$isoformFeatures %>% filter(condition_1 == "K20d", condition_2 == "K20w", gene_id == "AT2G36390")
test
#ok, so if we pull filter by condition_1, condition_2 and gene_id, we'll get the right rows. 
#for geneExp, we plot gene_value_1 and gene_value_2, plus or minus gene_stderr_1 and gene_stderr_2
#for isoExp, we plot iso_value_1 and iso_value_2, plus or minus iso_stderr_1 and iso_stderr_2
#for isoUsage, we plot iso_value_1 over gene-value_1, and iso_value_2 over gene_value_2, plus or minus iso_stderr_1 over gene_value_1 and iso_stderr_2 over gene-value_2
#the only q-value we have is for isoform usage, but that's the main graph anyway
#but seeing as I'm never going to be able to justify that, it would be best to run an in-house test to see if isoform usage is different between treatments and annotate significance that way
#can we do pairwise t-test and apply a Bonferroni correction, or similar?
#and ffs make the axis labels bigger...
```

